featurebyte/service/feature_table_cache.py

```diff
@@ -1,111 +1,367 @@
+            expressions.alias_(
+                node.name,
+                alias=f"FEATURE_{definition_hash}",
+                quoted=True,
+            )
+            for definition_hash, node in nodes.items()
+        ]
+        adapter = get_sql_adapter(db_session.source_type)
+        query = sql_to_string(
+            adapter.create_table_as(
+                table_details=TableDetails(
+                    database_name=db_session.database_name,
+                    schema_name=db_session.schema_name,
+                    table_name=final_table_name,
+                ),
+                select_expr=(
+                    expressions.select(InternalName.TABLE_ROW_INDEX)
+                    .select(*feature_names)
+                    .from_(quoted_identifier(intermediate_table_name))
+                ),
+            ),
+            source_type=db_session.source_type,
+        )
+        await db_session.execute_query(query)
+        await db_session.drop_table(
+            database_name=db_session.database_name,
+            schema_name=db_session.schema_name,
+            table_name=intermediate_table_name,
+            if_exists=True,
+        )
 
-        query_filter = {"observation_table_id": observation_table_id}
-        async for document in self.list_documents_iterator(query_filter=query_filter):
-            documents.append(document)
+    async def _update_table(
+        self,
+        feature_store: FeatureStoreModel,
+        observation_table: ObservationTableModel,
+        cache_metadata: FeatureTableCacheMetadataModel,
+        db_session: BaseSession,
+        graph: QueryGraph,
+        cached_nodes: Dict[str, Node],
+        non_cached_nodes: Dict[str, Node],
+        serving_names_mapping: Optional[Dict[str, str]] = None,
+        is_feature_list_deployed: bool = False,
+    ) -> None:
+        # create temporary table with features
+        temp_final_cache_table_name = (
+            f"__TEMP__{MaterializedTableNamePrefix.FEATURE_TABLE_CACHE}_{ObjectId()}"
+        )
+        await self._create_table(
+            feature_store=feature_store,
+            observation_table=observation_table,
+            db_session=db_session,
+            final_table_name=temp_final_cache_table_name,
+            graph=graph,
+            nodes=non_cached_nodes,
+            serving_names_mapping=serving_names_mapping,
+            is_feature_list_deployed=is_feature_list_deployed,
+        )
 
-        if not documents:
-            return None
+        # alter cached tables adding columns for new features
+        new_columns = [f"FEATURE_{definition_hash}" for definition_hash in non_cached_nodes.keys()]
+        alter_expr = expressions.AlterTable(
+            this=expressions.Table(this=quoted_identifier(cache_metadata.table_name)),
+            actions=[
+                expressions.ColumnDef(
+                    this=quoted_identifier(column),
+                    kind=expressions.DataType.build("?"),
```

Comment:
This function can be used to determine the dtype of the feature: https://github.com/featurebyte/featurebyte/blob/6ddfecc1ab9c0ba8f01737eea661e46ac84d7500/featurebyte/query_graph/transform/offline_store_ingest.py#L73-L75

Then we need to use adapter's `get_physical_type_from_dtype` method to convert it to the DataType here.